### ISSUE TITLE
[HOLD for payment 2024-06-06] [Track Tax] Add tax fields for distance requests

### ISSUE THREAD

**Issue by MonilBhavsar** [Expensify/App#31673]
## Problem

Internal Master tracking issue https://github.com/Expensify/Expensify/issues/338220
Coming from this section of design doc https://docs.google.com/document/d/1nkusVPiYN3uYCJAH_2XknBv_oPpNOifVxheuxRvFTxs/edit?pli=1#bookmark=id.d3mz10ifyyp0

We need to support tax tracking in distance requests also

### Frontend

1. We need to reuse the tax rate and tax amount components and also display it for Distance requests[Done]
2. Display tax fields if tax tracking on policy is enabled(`policy.tax.trackingEnabled` is true) **AND** tax tracking for distance requests is enabled(`policy.customUnits.attributes.taxEnabled` is true)
3. Both the tax rate and tax amount fields will be **uneditable** for distance requests
4. Ensure tax tracking works offline whenever we update route 

## Why is this important?
To allow users to track tax on newDot and migrate customers to newDot

## Solution
Implement the section of design doc

<details><summary>Issue Owner</summary>Current Issue Owner: @kevinksullivan</details>

**Comment by waterim** [comment]
Hey, Im Artem from Callstack and would like to take this issue!

**Comment by MonilBhavsar** [comment]
On HOLD

**Comment by kevinksullivan** [comment]
Still held

**Comment by kevinksullivan** [comment]
I also don't think I need to be here, so I'm unassigning. Let me know if I misunderstand this!

**Comment by MonilBhavsar** [comment]
Sounds good, I can assign you later if required 👍

**Comment by MonilBhavsar** [comment]
On HOLD

**Comment by MonilBhavsar** [comment]
Still on HOLD

**Comment by MonilBhavsar** [comment]
On HOLD

**Comment by waterim** [comment]
@MonilBhavsar Hello!
As I understand I can proceed with this issue as the main one was merged?

**Comment by MonilBhavsar** [comment]
@waterim only creation flow for manual request was merged. The PR to edit and view is still in review. Once it is merged, this issue will be unblocked. Thanks for checking!

**Comment by MonilBhavsar** [comment]
On HOLD

**Comment by MonilBhavsar** [comment]
on HOLD

**Comment by MonilBhavsar** [comment]
Still on HOLD

**Comment by MonilBhavsar** [comment]
on hold but resuming soon

**Comment by MonilBhavsar** [comment]
Taking this off HOLD and we can start with this cc @waterim

**Comment by waterim** [comment]
@MonilBhavsar Hello Monil, thank you!
@teneeto will proceed with this ticket if you don't mind :)

**Comment by teneeto** [comment]
Sure I can, @waterim

**Comment by MonilBhavsar** [comment]
Yes for sure! Assigning @teneeto to the issue

**Comment by MonilBhavsar** [comment]
Shifted to other wave issue, circling back to this

**Comment by MonilBhavsar** [comment]
Backend PR is in review!

**Comment by MonilBhavsar** [comment]
@teneeto this is ready to be worked on!

**Comment by MonilBhavsar** [comment]
We're fixing some bugs and will circle back

**Comment by MonilBhavsar** [comment]
PR for supporting updating requests is in review

**Comment by Beamanator** [comment]
@MonilBhavsar just curious, do you think I'll have to deal with tax updates for the new API i'm building, `UpdateDistanceRequestRate`, here? https://github.com/Expensify/App/issues/36969

**Comment by MonilBhavsar** [comment]
Hey! From a quick look, it's for P2P requests that are sent from one user to another and not associated with workspace, right?
If so, no we don't need to include tax for P2P requests because tax is only for workspace expense chats and for workspace who have tax tracking enabled

Please see this line https://github.com/Expensify/Web-Expensify/blob/d195dfa247c2e6a4218f417d8e1166f486f91cb1/lib/TransactionAPI.php#L1880

We return early if it's not workspace chat or tax tracking is disabled, which I believe would be the case for P2P

**Comment by Beamanator** [comment]
Mmm good call! I believe the main project goal is for P2P, but this command MAY be even used for money requests in workspace chats - maybe @neil-marcellini can help there, but I'll keep that in mind for my PR, thanks!!

**Comment by MonilBhavsar** [comment]
Ah okay! Feel free to tag me for review

**Comment by Beamanator** [comment]
Will do, if I end up touching the tax shtuff you wrote :D 🙏  thanks!

**Comment by neil-marcellini** [comment]
`UpdateMoneyRequestRate` will only be used for updating rates of workspace distance requests. It's not possible to edit the rate of P2P distance requests because they always use a hard-coded default value based on the personal policy currency.

**Comment by Beamanator** [comment]
ok ok ok so i was very wrong 😅 

so we DO need to deal with tax stuff, AND i can't use what @MonilBhavsar wrote (most likely) since I'm going to be doing most of my changes in Auth for 1:1:1 sake 😅

**Comment by MonilBhavsar** [comment]
You can't use now, but should be able to use soon. I am also moving some stuff to Auth for another PR https://github.com/Expensify/Auth/pull/10517/files#diff-dccb44cf036f0d8f9ea7278bacf5670b5e6b5d1f11f48fc203644400c1806d15R3088 
We co ordinate and make it 1:1:1 for all cases 🤝

**Comment by Beamanator** [comment]
Ooh that's amazing!!! I wrote a plan ([here](https://github.com/Expensify/App/issues/36969#issuecomment-2056543866)) for how I plan to move forward making `UpdateMoneyRequestRate` 1:1:1, and I'll note that I'll need to incorporate your upcoming Auth Tax stuff as well 🙏

**Comment by MonilBhavsar** [comment]
Backend PR's are deployed. Frontend PR will up soon, once https://github.com/Expensify/App/pull/40443 is merged

**Comment by teneeto** [comment]
@MonilBhavsar, If both tax rate and tax amount values are uneditable for  distance requests, does it mean tax rate for distance requests will always be set to default?

**Comment by teneeto** [comment]
@MonilBhavsar a draft for this is [here](https://github.com/Expensify/App/pull/41379)

**Comment by MonilBhavsar** [comment]
> If both tax rate and tax amount values are uneditable for distance requests, does it mean tax rate for distance requests will always be set to default?

Good question! Previously it would have remain same. But not anymore, because we recently introduced a field to select the distance rate

<img width="1722" alt="Screenshot 2024-05-01 at 12 37 14 PM" src="https://github.com/Expensify/App/assets/32012005/c84c365d-7437-4832-b86f-6d98acbdd98a">

Every tax rate has a taxCode associated with it. So, we need to update the taxRate when the distance rate changes.
Whenever distance track tax is enabled, `policy.customUnits.rates.attributes.taxExternalID`

**Comment by MonilBhavsar** [comment]
Also, if distance rate updates, tax rate will be updated and hence tax amount too.

The formulae to calculate tax for distance amount is slightly different

```js
const taxClaimablePercentage = policy.customUnits.rates.attributes.taxClaimablePercentage;
const taxableAmount = -1 * expenseAmount* taxClaimablePercentage;
const taxPercentage = getTaxValue();
const taxAmount = TransactionUtils::calculateTaxAmount(taxPercentage, taxableAmount);
 ```

**Comment by teneeto** [comment]
In the shared screenshot, I see 2 rates, in that case which of them is the rate to be used? they both seem to be enabled.

**Comment by teneeto** [comment]
> Also, if distance rate updates, tax rate will be updated and hence tax amount too.
> 
> The formulae to calculate tax for distance amount is slightly different
> 
> ```js
> const taxClaimablePercentage = policy.customUnits.rates.attributes.taxClaimablePercentage;
> const taxableAmount = -1 * expenseAmount* taxClaimablePercentage;
> const taxPercentage = getTaxValue();
> const taxAmount = TransactionUtils::calculateTaxAmount(taxPercentage, taxableAmount);
> ```

how will i tell a distance rate is updated?

**Comment by MonilBhavsar** [comment]
@neil-marcellini do we not allow updating distance rate when editing request?

**Comment by neil-marcellini** [comment]
> @neil-marcellini do we not allow updating distance rate when editing request?

Right now no, we do not have a way to update the distance rate. We are planning to allow updating the distance rate for workspace requests with this issue [[P2P Distance] Create UpdateMoneyRequestDistanceRate](https://github.com/Expensify/App/issues/36969). It will not be allowed for P2P requests in the foreseeable future.

**Comment by MonilBhavsar** [comment]
@teneeto I believe you're working on a PR. Anything you want to add here?

**Comment by teneeto** [comment]
@MonilBhavsar Yes, I am working on it. No questions yet, let me address the new changes.

**Comment by MonilBhavsar** [comment]
Being worked on ^

**Comment by MonilBhavsar** [comment]
Latest is here https://expensify.slack.com/archives/C036QM0SLJK/p1715597553123279?thread_ts=1715336572.603089&cid=C036QM0SLJK

**Comment by narefyev91** [comment]
Hey - I'm Nicolay from Callstack - will continue here until Eto will be off

**Comment by MonilBhavsar** [comment]
@kevinksullivan can we please process payment for @ishpaul777 for C+ role

**Comment by kevinksullivan** [comment]
job is here

https://www.upwork.com/jobs/~010d7783c2efb71b9d

**Comment by kevinksullivan** [comment]
Offer sent to @ishpaul777 . @teneeto I cannot find you on upwork, mind applying to the posting directly?

**Comment by kevinksullivan** [comment]
Paid @ishpaul777 !

**Comment by MonilBhavsar** [comment]
Thank you kevin! Going to close this one out
